### ISSUE TITLE
Collaboration Cycle for [10-10 Health Apps, 10-10CG, Improve Facility Selection]

### ISSUE THREAD

**Issue by hdjustice** [department-of-veterans-affairs/va.gov-team#51980]
### If your product is already on VA.gov, please verify your product information in the Product Directory.

- [X] The product information in the Product Directory is accurate.

### VFS team name

10-10 Health Apps

### Product name

10-10CG

### Feature name

Improve Facility Selection

### GitHub label for product

1010-cg

### GitHub label for feature

cg-facilities

### When did/will you start working on this product/feature?

The feature was initially started in Feb 2021, however it became stale and outdated since then.  We have picked up discussions since beginning of December 2022 to revisit the design and purpose of the page/question.

### Will your work result in visible changes to the user experience?

Yes

### Are you doing research with VA.gov users?

Yes

### Will your work involve changes to...

Tools, applications and dynamic pages

### Does your product/feature have Google Analytics tracking and a KPI dashboard in Domo?

Yes

### Do you need to capture any additional analytics or metrics?

No

### Product outline

https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/caregivers/10-10CG%20Caregiver%20application%20product-outline.md

### Notify the Collaboration Cycle team of this ticket in the vfs-platform-support Slack channel.

- [X] I acknowledge that I must notify \#vfs-platform-support after submitting this issue.

**Comment by JonathanKamensDVA** [comment]
@hdjustice I do not see the arch intent documents linked above. Can you please add links above in the "Architecture Intent artifacts for review" section above and comment here when that's finished?

**Comment by hdjustice** [comment]
@JonathanKamensDVA  We haven't finished them yet but we are in the process of working on the checklist. We are not quite ready to schedule the meeting but as soon as we are, we will follow the instructions and comment here.  Thanks!!

**Comment by hdjustice** [comment]
@JonathanKamensDVA we have the docs linked for Architecture Intent, scheduled for next Tuesday.  I apologize for the late comment, I did everything but that

**Comment by hdjustice** [comment]
@JonathanKamensDVA (and any other Architecture Intent folks)

**Comment by hdjustice** [comment]
@JonathanKamensDVA (and any other Architecture Intent folks)
We discussed the GET to POST change we talked about during the review meeting and here is the option I think we want to take. Can this be reviewed and make sure we are not causing any other issues or concerns?

Add ability to encrypt/decrypt sensitive query params as needed. I took a crack at spiking this out and it is certainly doable. I used RSA encryption with the [WebCrypto api](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) and have [this vets-website spike branch](https://github.com/department-of-veterans-affairs/vets-website/tree/98100-encrypt-query-params) in conjunction with [this vets-api branch to decrypt the params on the backend](https://github.com/department-of-veterans-affairs/vets-api/tree/98100-encrypt-query-params) that is using OpenSSL on the backend to decrypt it. This would take the most work to implement, and would involve new public/private keys and the subsequent configuration of them. One advantage is it could easily be implemented to allow for other teams to use, which could be a big benefit. It would allow us to keep the request as a GET, which would keep the api RESTful.

cc: @coope93

**Comment by JonathanKamensDVA** [comment]
Hi @hdjustice! I just want to make sure I'm understanding the proposed solution properly...

Problem statement: Eliminate or obscure PII in URL query parameters, because URLs get logged and recorded all over the place so if there's PII in them we essentially lose control over where the PII goes.

Proposed solution: Send a public key down to the browser and use it to encrypt query parameters that could potentially contain PII. Decrypt those query parameters on the server using a private key retrieved in memory from the secrets store. Assuming that a sufficiently strong keypair is used, this ensures that the data in the query parameters are effectively inaccessible to anyone who has access to any of the datastores in which URLs are logged and recorded.

If I'm understanding correctly, then here's my take on this solution...

It is definitely a major improvement.

However, it is not ideal, because any insider with access to the secrets store (or attacker who steals access to the secrets store) can pull the private key from it and use it to decrypt any of the PII in the URLs.

I would suggest mitigating that risk by frequently rotating the keypair. The rotation should be automatic (perhaps part of the deployment process?) so that we can effectively do it as frequently as every day without any manual overhead. You'd need to be careful about how this is done to ensure that people who are using the site while the rotation is happening don't suddenly have an action fail because the key was rotated between when they loaded a page and when they submitted a query from it. Something like this:

* A primary and secondary key are supported. The backend tries to decrypt using the primary key first, and if that fails, tries the secondary key. We only store the private-key half of the secondary key.
* Each deployment rotates the key if it was last rotated more than 12 hours ago:
   * Copy the private half of the primary key to the secondary key.
   * Generate and deploy a new primary key (both public and private).
* The front-end reloads the public key from the secrets store at least hourly.

What do you think?

**Comment by JonathanKamensDVA** [comment]
@DonMcCaugheyUSDS @JeffKeeneVAGov @kaipyroami now would be the time to speak up about ⏫ if you think I'm being ridiculously paranoid here. 😉

**Comment by coope93** [comment]
@JonathanKamensDVA That makes sense, although it definitely broadens the scope a bit. Just to confirm, using a `POST` call with the latitude and longitude in the request body would also be a valid solution, correct? While I understand it’s not RESTful, I’ve noticed we handle it that way in other places in `vets-api`—specifically, the existing Facility Locator uses this approach.

While reviewing the new facility search code, I came across another point I’d like your input on. Currently, we make a `GET` request to Mapbox from `vets-website`, passing the user-entered search text as query parameters. This request retrieves the latitude and longitude, which is then sent to vets-api (where the data is supposed to be encrypted). This Mapbox call is used in other places within `vets-website`—for example, the VA Facility Locator and a feature in Ask VA. Our implementation is based on those patterns.

Since this seems closely related to your feedback about sensitive information in query parameters, how do you recommend we address this? It feels like a similar concern, but I’m unsure how we could resolve it without significant changes, such as reworking our approach or moving away from Mapbox entirely.

**Comment by JonathanKamensDVA** [comment]
> [@JonathanKamensDVA](https://github.com/JonathanKamensDVA) That makes sense, although it definitely broadens the scope a bit.

Yeah, that's why I asked Don, Jeff, and Kyle to talk me down if they think I'm being overly paranoid. 😉 If we're going to build something like this with the intention of other teams being able to use it, then we can assume that the amount of PII being encrypted as per this proposal will grow to a significant size over time, so I do think we have to be careful to get it right from the start.

> Just to confirm, using a `POST` call with the latitude and longitude in the request body would also be a valid solution, correct?

Yes, correct.

> While reviewing the new facility search code, I came across another point I’d like your input on. Currently, we make a `GET` request to Mapbox from `vets-website`, passing the user-entered search text as query parameters.

This has come up before.

The current implementation is not ideal because this means that we are giving Mapbox the user's IP address along with information about a geographic location they care about, so that's two disparate pieces of PII about the same individual.

It would be preferable for us to be proxying the Mapbox calls through vets-api so that the IP address that Mapbox sees is the vets-api egress IP address, not the user's IP address. Then we would only be telling Mapbox, "Some user of VA.gov is interested in information about this geographic location," grouped together with Mapbox queries about a bunch of _other_ geographic locations, and therefore essentially aggregated and not substantive PII.

Having said that, as you point out, we're already doing this wrong elsewhere in the app, so you're not creating a new problem by following the same pattern in your app that we're already following elsewhere.

I mean, if you _want_ to build a new vets-api endpoint to proxy the Mapbox calls, which other teams can then leverage to ensure appropriate geo privacy for their Mapbox queries, I certainly won't object. At the same time, I'm not going to burden your team with solving a problem that's already in the code base. So until we decide as an org that we need to invest effort into fixing this issue, you're not on the hook to fix it yourselves.

**Comment by DonMcCaugheyUSDS** [comment]
Sorry I'm late in commenting on this.

1.) If I understand correctly from the conversation above and from looking at the arch intent checklist and Figma file, we ask for the general location of a VA facility ("What VA medical center or clinic..." -> "City, state or postal code").  This doesn't seem like PII to me.  @JonathanKamensDVA, am I missing something?

2.) It seems we connect directly to Mapbox from vets-website and thus send the user's IP address to Mapbox.  This is more concerning to me, as we are effectively sharing PII with Mapbox.  I hope we have an account with them and thus some level of assurance that they're properly handling this PII, but I agree with Jonathan that proxying Mapbox calls through vets-api is preferable.  Such a proxy endpoint seems like it would be straightforward to implement; it's stateless and cacheable.

3.) Ideally an HTTP search endpoint should be an idempotent GET with URL parameters.  GETs are cacheable, bookmark-able and easily shareable.  But once you add private data to query params, cache-ability isn't so important and bookmarking & sharing becomes problematic.  Just use a POST with the query params in the body.  I feel like any encryption scheme for URL parameters is going to cause headaches in the future.  Is there any reason not to use a POST here?

**Comment by JonathanKamensDVA** [comment]
> 1.) If I understand correctly from the conversation above and from looking at the arch intent checklist and Figma file, we ask for the general location of a VA facility ("What VA medical center or clinic..." -> "City, state or postal code"). This doesn't seem like PII to me. [@JonathanKamensDVA](https://github.com/JonathanKamensDVA), am I missing something?

A few things. 😉

1. The description does say "City, state or postal code," but it's a freetext field and the user can enter anything there that can be parsed into a location, including their street address. I am _certain_ that some users do that.
2. In fact, if the user clicks "Use my location" and gives the browser location access when prompted to do so, their precise location&mdash;as precise as the browser is able to obtain&mdash;is inserted automatically into the text field. I just tried it on my phone and it inserted the full street address of my home into the text field.
3. City/State and ZIP code are both considered PII when combined with other information about an individual. It doesn't take an exact street address to be PII.
